File: 0018 4sum/solution.go
func fourSum(nums []int, target int) [][]int {
    sort.Ints(nums);
    n:=len(nums)
    var ans [][]int
    for k:=0;k<n-3;k++ {
        if k>0 && nums[k]==nums[k-1] {
            continue
        }
        for i:=k+1;i<n-2;i++ {
            if i>k+1 && nums[i]==nums[i-1] {
                continue
            }
            l,r:=i+1,n-1
            for l<r {
                sum:=nums[k]+nums[i]+nums[l]+nums[r]
                if sum==target {
                    mat:=[]int{nums[k],nums[i],nums[l],nums[r]}
                    ans = append(ans,mat)
                    for l<r && nums[l]==nums[l+1] {
                        l++
                    }
                    for l<r && nums[r]==nums[r-1] {
                        r--
                    }
                    l++
                    r--
                } else if sum < target {
                    l++
                } else {
                    r--
                }
            }
        }
    }
    return ans
}

File: 0018 4sum/solution.ts
function fourSum(nums: number[], target: number): number[][] {
    let ans: number[][] = [];
    nums.sort((a, b) => a - b);
    for(let i=0;i<nums.length-3;i++) {
     if(i > 0 && nums[i]===nums[i-1]) continue;
     for(let k=i+1;k<nums.length-2;k++) {
        if(k > i+1 && nums[k]===nums[k-1]) continue;
        let l = k+1,r=nums.length-1;
        while(l<r) {
            let sum = nums[k]+nums[i]+nums[l]+nums[r];
            if(sum==target) {
                ans.push([nums[i], nums[k] ,nums[l], nums[r]]);
                while(l<r && nums[l]==nums[l+1]) l++;
                while(l<r && nums[r]==nums[r-1]) r--;
                l++;
                r--;
            }
            else if(sum < target) l++;
            else r--; 
        }
    }
    }
    return ans;
};
